File: frontend/src/services/LocalizationService.ts
/**
 * AisleMarts Localization Service - Frontend
 * Handles automatic localization, currency display, and language preferences
 */

import AsyncStorage from '@react-native-async-storage/async-storage';
import axios from 'axios';
import Constants from 'expo-constants';

// Get backend URL from env
const BACKEND_URL = Constants.expoConfig?.extra?.EXPO_BACKEND_URL || 'http://localhost:8001';

export interface LocalizationData {
  country_code: string;
  country_name: string;
  currency: string;
  currency_symbol: string;
  currency_name: string;
  language: string;
  city?: string;
  region?: string;
  ip: string;
}

export interface CurrencyConversion {
  amount: number;
  currency: string;
  symbol: string;
  formatted: string;
  conversion_rate: number;
  original_amount?: number;
  original_currency?: string;
}

export interface SupportedCountry {
  currency: string;
  symbol: string;
  name: string;
  language: string;
}

class LocalizationService {
  private localizationData: LocalizationData | null = null;
  private supportedCountries: Record<string, SupportedCountry> = {};
  
  // Cache keys
  private readonly LOCALIZATION_CACHE_KEY = '@aislemarts_localization';
  private readonly CURRENCY_PREFERENCE_KEY = '@aislemarts_currency_preference';
  private readonly LANGUAGE_PREFERENCE_KEY = '@aislemarts_language_preference';
  
  constructor() {
    this.loadCachedData();
  }

  /**
   * Load cached localization data and preferences
   */
  private async loadCachedData(): Promise<void> {
    try {
      const cachedData = await AsyncStorage.getItem(this.LOCALIZATION_CACHE_KEY);
      if (cachedData) {
        this.localizationData = JSON.parse(cachedData);
      }
    } catch (error) {
      console.error('Error loading cached localization data:', error);
    }
  }

  /**
   * Auto-detect user's location and preferences
   */
  async detectUserLocalization(): Promise<LocalizationData> {
    try {
      const response = await axios.get(`${BACKEND_URL}/api/localization/detect`);
      
      if (response.data) {
        this.localizationData = response.data;
        
        // Cache the data
        await AsyncStorage.setItem(
          this.LOCALIZATION_CACHE_KEY, 
          JSON.stringify(this.localizationData)
        );
        
        console.log('Localization detected:', this.localizationData);
        return this.localizationData;
      }
    } catch (error) {
      console.error('Error detecting localization:', error);
    }
    
    // Fallback to Kenya (pilot market)
    const fallbackData: LocalizationData = {
      country_code: 'KE',
      country_name: 'Kenya',
      currency: 'KES',
      currency_symbol: 'KSh',
      currency_name: 'Kenyan Shilling',
      language: 'en',
      ip: 'unknown'
    };
    
    this.localizationData = fallbackData;
    return fallbackData;
  }

  /**
   * Get current localization data (cached or detect if not available)
   */
  async getCurrentLocalization(): Promise<LocalizationData> {
    if (this.localizationData) {
      return this.localizationData;
    }
    
    return await this.detectUserLocalization();
  }

  /**
   * Convert price to user's preferred currency
   */
  async convertPrice(
    amount: number, 
    fromCurrency: string, 
    toCurrency?: string
  ): Promise<CurrencyConversion> {
    try {
      // Use user's current currency if not specified
      if (!toCurrency) {
        const localization = await this.getCurrentLocalization();
        toCurrency = await this.getPreferredCurrency() || localization.currency;
      }
      
      const response = await axios.post(`${BACKEND_URL}/api/localization/convert-currency`, {
        amount,
        from_currency: fromCurrency,
        to_currency: toCurrency
      });
      
      return response.data;
    } catch (error) {
      console.error('Error converting currency:', error);
      
      // Fallback conversion (basic estimation)
      const localization = await this.getCurrentLocalization();
      return {
        amount: amount,
        currency: localization.currency,
        symbol: localization.currency_symbol,
        formatted: `${localization.currency_symbol}${amount.toFixed(2)}`,
        conversion_rate: 1.0
      };
    }
  }

  /**
   * Format price in user's currency
   */
  async formatPrice(amount: number, currency?: string): Promise<string> {
    try {
      const localization = await this.getCurrentLocalization();
      const targetCurrency = currency || localization.currency;
      
      if (targetCurrency === localization.currency) {
        return `${localization.currency_symbol}${amount.toLocaleString('en-US', {
          minimumFractionDigits: 2,
          maximumFractionDigits: 2
        })}`;
      }
      
      // Convert if different currency
      const conversion = await this.convertPrice(amount, 'USD', targetCurrency);
      return conversion.formatted;
    } catch (error) {
      console.error('Error formatting price:', error);
      return `$${amount.toFixed(2)}`;
    }
  }

  /**
   * Get localized greeting for Aisle AI
   */
  async getLocalizedGreeting(): Promise<string> {
    try {
      const localization = await this.getCurrentLocalization();
      const language = await this.getPreferredLanguage() || localization.language;
      
      const response = await axios.get(
        `${BACKEND_URL}/api/localization/greeting/${localization.country_code}?language=${language}`
      );
      
      return response.data.greeting;
    } catch (error) {
      console.error('Error getting localized greeting:', error);
      return "Welcome to AisleMarts! 🌍 I'm Aisle, your global shopping companion ready to help you discover amazing products worldwide.";
    }
  }

  /**
   * Get supported countries
   */
  async getSupportedCountries(): Promise<Record<string, SupportedCountry>> {
    try {
      if (Object.keys(this.supportedCountries).length === 0) {
        const response = await axios.get(`${BACKEND_URL}/api/localization/supported-countries`);
        this.supportedCountries = response.data.countries;
      }
      
      return this.supportedCountries;
    } catch (error) {
      console.error('Error getting supported countries:', error);
      return {};
    }
  }

  /**
   * Set user's preferred currency
   */
  async setPreferredCurrency(currency: string): Promise<void> {
    try {
      await AsyncStorage.setItem(this.CURRENCY_PREFERENCE_KEY, currency);
      console.log('Currency preference set:', currency);
    } catch (error) {
      console.error('Error setting currency preference:', error);
    }
  }

  /**
   * Get user's preferred currency
   */
  async getPreferredCurrency(): Promise<string | null> {
    try {
      return await AsyncStorage.getItem(this.CURRENCY_PREFERENCE_KEY);
    } catch (error) {
      console.error('Error getting currency preference:', error);
      return null;
    }
  }

  /**
   * Set user's preferred language
   */
  async setPreferredLanguage(language: string): Promise<void> {
    try {
      await AsyncStorage.setItem(this.LANGUAGE_PREFERENCE_KEY, language);
      console.log('Language preference set:', language);
    } catch (error) {
      console.error('Error setting language preference:', error);
    }
  }

  /**
   * Get user's preferred language
   */
  async getPreferredLanguage(): Promise<string | null> {
    try {
      return await AsyncStorage.getItem(this.LANGUAGE_PREFERENCE_KEY);
    } catch (error) {
      console.error('Error getting language preference:', error);
      return null;
    }
  }

  /**
   * Check if country is supported
   */
  async isCountrySupported(countryCode: string): Promise<boolean> {
    const countries = await this.getSupportedCountries();
    return countryCode in countries;
  }

  /**
   * Get exchange rate between two currencies
   */
  async getExchangeRate(fromCurrency: string, toCurrency: string): Promise<number> {
    try {
      const response = await axios.get(
        `${BACKEND_URL}/api/localization/exchange-rate/${fromCurrency}/${toCurrency}`
      );
      
      return response.data.rate;
    } catch (error) {
      console.error('Error getting exchange rate:', error);
      return 1.0;
    }
  }

  /**
   * Get currency info
   */
  async getCurrencyInfo(currencyCode: string): Promise<any> {
    try {
      const response = await axios.get(
        `${BACKEND_URL}/api/localization/currency-info/${currencyCode}`
      );
      
      return response.data;
    } catch (error) {
      console.error('Error getting currency info:', error);
      return {
        currency_code: currencyCode,
        currency: currencyCode,
        symbol: '$',
        name: 'Unknown Currency'
      };
    }
  }

  /**
   * Clear cached data (useful for testing or user logout)
   */
  async clearCache(): Promise<void> {
    try {
      await AsyncStorage.multiRemove([
        this.LOCALIZATION_CACHE_KEY,
        this.CURRENCY_PREFERENCE_KEY,
        this.LANGUAGE_PREFERENCE_KEY
      ]);
      
      this.localizationData = null;
      this.supportedCountries = {};
      
      console.log('Localization cache cleared');
    } catch (error) {
      console.error('Error clearing localization cache:', error);
    }
  }

  /**
   * Get country flag emoji
   */
  getCountryFlag(countryCode: string): string {
    const flagMap: Record<string, string> = {
      'KE': '🇰🇪', 'NG': '🇳🇬', 'ZA': '🇿🇦', 'EG': '🇪🇬', 'MA': '🇲🇦',
      'IT': '🇮🇹', 'DE': '🇩🇪', 'FR': '🇫🇷', 'ES': '🇪🇸', 'GB': '🇬🇧',
      'TR': '🇹🇷', 'US': '🇺🇸', 'CA': '🇨🇦', 'MX': '🇲🇽', 'BR': '🇧🇷',
      'IN': '🇮🇳', 'CN': '🇨🇳', 'JP': '🇯🇵', 'AE': '🇦🇪', 'SA': '🇸🇦'
    };
    
    return flagMap[countryCode] || '🌍';
  }

  /**
   * Get localized AisleMarts branding
   */
  async getLocalizedBranding(): Promise<{title: string, subtitle: string}> {
    const localization = await this.getCurrentLocalization();
    const flag = this.getCountryFlag(localization.country_code);
    
    return {
      title: `AisleMarts ${localization.country_name}`,
      subtitle: `Global Marketplace • Local Power ${flag}`
    };
  }
}

// Export singleton instance
export const localizationService = new LocalizationService();
export default localizationService;

File: backend/localization_service.py
"""
AisleMarts Localization Service - Auto-detect user location and provide localized experience
Supports currency detection, language preferences, and regional customization
"""

import requests
import asyncio
from typing import Dict, Optional, Tuple
from datetime import datetime, timedelta
from fastapi import Request
import json
import logging
from dotenv import load_dotenv
import os

load_dotenv()

logger = logging.getLogger(__name__)

class LocalizationService:
    """
    Handles automatic localization for AisleMarts users including:
    - IP-based location detection
    - Currency assignment based on country
    - Language preference detection
    - Regional product filtering
    """
    
    def __init__(self):
        # Currency mapping for major AisleMarts target markets
        self.country_currency_map = {
            # Africa
            'KE': {'currency': 'KES', 'symbol': 'KSh', 'name': 'Kenyan Shilling'},
            'NG': {'currency': 'NGN', 'symbol': '₦', 'name': 'Nigerian Naira'},
            'ZA': {'currency': 'ZAR', 'symbol': 'R', 'name': 'South African Rand'},
            'EG': {'currency': 'EGP', 'symbol': 'LE', 'name': 'Egyptian Pound'},
            'MA': {'currency': 'MAD', 'symbol': 'MAD', 'name': 'Moroccan Dirham'},
            
            # Europe
            'IT': {'currency': 'EUR', 'symbol': '€', 'name': 'Euro'},
            'DE': {'currency': 'EUR', 'symbol': '€', 'name': 'Euro'},
            'FR': {'currency': 'EUR', 'symbol': '€', 'name': 'Euro'},
            'ES': {'currency': 'EUR', 'symbol': '€', 'name': 'Euro'},
            'GB': {'currency': 'GBP', 'symbol': '£', 'name': 'British Pound'},
            'TR': {'currency': 'TRY', 'symbol': '₺', 'name': 'Turkish Lira'},
            
            # Americas
            'US': {'currency': 'USD', 'symbol': '$', 'name': 'US Dollar'},
            'CA': {'currency': 'CAD', 'symbol': 'C$', 'name': 'Canadian Dollar'},
            'MX': {'currency': 'MXN', 'symbol': '$', 'name': 'Mexican Peso'},
            'BR': {'currency': 'BRL', 'symbol': 'R$', 'name': 'Brazilian Real'},
            
            # Asia
            'IN': {'currency': 'INR', 'symbol': '₹', 'name': 'Indian Rupee'},
            'CN': {'currency': 'CNY', 'symbol': '¥', 'name': 'Chinese Yuan'},
            'JP': {'currency': 'JPY', 'symbol': '¥', 'name': 'Japanese Yen'},
            'AE': {'currency': 'AED', 'symbol': 'AED', 'name': 'UAE Dirham'},
            'SA': {'currency': 'SAR', 'symbol': 'SAR', 'name': 'Saudi Riyal'},
        }
        
        # Language mapping for countries
        self.country_language_map = {
            'KE': 'en',  # English/Swahili
            'NG': 'en',  # English
            'ZA': 'en',  # English
            'EG': 'ar',  # Arabic
            'MA': 'ar',  # Arabic/French
            'IT': 'it',  # Italian
            'DE': 'de',  # German
            'FR': 'fr',  # French
            'ES': 'es',  # Spanish
            'GB': 'en',  # English
            'TR': 'tr',  # Turkish
            'US': 'en',  # English
            'CA': 'en',  # English/French
            'MX': 'es',  # Spanish
            'BR': 'pt',  # Portuguese
            'IN': 'en',  # English/Hindi
            'CN': 'zh',  # Chinese
            'JP': 'ja',  # Japanese
            'AE': 'ar',  # Arabic
            'SA': 'ar',  # Arabic
        }
        
        # Free IP geolocation service
        self.ip_api_url = "http://ip-api.com/json/"
        
        # Currency conversion cache
        self.currency_cache = {}
        self.cache_expiry = {}
        
    async def detect_user_location(self, request: Request) -> Dict[str, str]:
        """
        Detect user location from IP address
        Returns country code, currency, and language
        """
        try:
            # Get IP address from request
            client_ip = self.get_client_ip(request)
            
            # For local development, return default
            if client_ip in ['127.0.0.1', 'localhost', '::1']:
                return {
                    'country_code': 'KE',  # Default to Kenya for testing
                    'country_name': 'Kenya',
                    'currency': 'KES',
                    'currency_symbol': 'KSh',
                    'language': 'en',
                    'ip': client_ip
                }
            
            # Call IP geolocation API
            response = requests.get(f"{self.ip_api_url}{client_ip}", timeout=5)
            
            if response.status_code == 200:
                location_data = response.json()
                
                if location_data.get('status') == 'success':
                    country_code = location_data.get('countryCode', 'US')
                    country_name = location_data.get('country', 'United States')
                    
                    # Get currency and language for this country
                    currency_info = self.country_currency_map.get(country_code, self.country_currency_map['US'])
                    language = self.country_language_map.get(country_code, 'en')
                    
                    return {
                        'country_code': country_code,
                        'country_name': country_name,
                        'currency': currency_info['currency'],
                        'currency_symbol': currency_info['symbol'],
                        'currency_name': currency_info['name'],
                        'language': language,
                        'city': location_data.get('city', ''),
                        'region': location_data.get('regionName', ''),
                        'ip': client_ip
                    }
            
        except Exception as e:
            logger.error(f"Error detecting user location: {str(e)}")
        
        # Fallback to default (Kenya for AisleMarts pilot)
        return {
            'country_code': 'KE',
            'country_name': 'Kenya',
            'currency': 'KES',
            'currency_symbol': 'KSh',
            'currency_name': 'Kenyan Shilling',
            'language': 'en',
            'ip': 'unknown'
        }
    
    def get_client_ip(self, request: Request) -> str:
        """Extract client IP from request headers"""
        # Check for forwarded IP (for load balancers/proxies)
        forwarded_for = request.headers.get('X-Forwarded-For')
        if forwarded_for:
            return forwarded_for.split(',')[0].strip()
        
        # Check for real IP header
        real_ip = request.headers.get('X-Real-IP')
        if real_ip:
            return real_ip
        
        # Fallback to direct connection IP
        return request.client.host if request.client else '127.0.0.1'
    
    async def get_currency_conversion_rate(self, from_currency: str, to_currency: str) -> float:
        """
        Get currency conversion rate from free API
        Returns conversion rate from from_currency to to_currency
        """
        if from_currency == to_currency:
            return 1.0
        
        cache_key = f"{from_currency}_{to_currency}"
        
        # Check cache (1-hour expiry)
        if cache_key in self.currency_cache:
            if datetime.now() < self.cache_expiry.get(cache_key, datetime.now()):
                return self.currency_cache[cache_key]
        
        try:
            # Using exchangerate-api.com (free tier: 1500 requests/month)
            api_url = f"https://api.exchangerate-api.com/v4/latest/{from_currency}"
            
            response = requests.get(api_url, timeout=5)
            
            if response.status_code == 200:
                data = response.json()
                rate = data.get('rates', {}).get(to_currency)
                
                if rate:
                    # Cache for 1 hour
                    self.currency_cache[cache_key] = rate
                    self.cache_expiry[cache_key] = datetime.now() + timedelta(hours=1)
                    return rate
            
        except Exception as e:
            logger.error(f"Error getting currency conversion rate: {str(e)}")
        
        # Fallback rates (approximate, for demo purposes)
        fallback_rates = {
            'USD_KES': 150.0,
            'USD_EUR': 0.85,
            'USD_GBP': 0.75,
            'USD_NGN': 800.0,
            'USD_EGP': 31.0,
            'USD_TRY': 28.0,
            'KES_USD': 0.0067,
            'EUR_USD': 1.18,
            'GBP_USD': 1.33,
        }
        
        return fallback_rates.get(cache_key, 1.0)
    
    async def convert_price(self, price: float, from_currency: str, to_currency: str) -> Dict[str, any]:
        """
        Convert price from one currency to another
        Returns converted price with formatting
        """
        if from_currency == to_currency:
            currency_info = self.get_currency_info(to_currency)
            return {
                'amount': price,
                'currency': to_currency,
                'symbol': currency_info['symbol'],
                'formatted': f"{currency_info['symbol']}{price:,.2f}",
                'conversion_rate': 1.0
            }
        
        # Get conversion rate
        rate = await self.get_currency_conversion_rate(from_currency, to_currency)
        converted_amount = price * rate
        
        # Get currency info for formatting
        currency_info = self.get_currency_info(to_currency)
        
        return {
            'amount': round(converted_amount, 2),
            'currency': to_currency,
            'symbol': currency_info['symbol'],
            'formatted': f"{currency_info['symbol']}{converted_amount:,.2f}",
            'conversion_rate': rate,
            'original_amount': price,
            'original_currency': from_currency
        }
    
    def get_currency_info(self, currency_code: str) -> Dict[str, str]:
        """Get currency information by currency code"""
        for country_currencies in self.country_currency_map.values():
            if country_currencies['currency'] == currency_code:
                return country_currencies
        
        # Fallback
        return {'currency': 'USD', 'symbol': '$', 'name': 'US Dollar'}
    
    def get_supported_countries(self) -> Dict[str, Dict]:
        """Get list of all supported countries with their currency/language info"""
        return {
            code: {
                **currency_info,
                'language': self.country_language_map.get(code, 'en')
            }
            for code, currency_info in self.country_currency_map.items()
        }
    
    def is_country_supported(self, country_code: str) -> bool:
        """Check if a country is supported by AisleMarts"""
        return country_code in self.country_currency_map
    
    async def get_localized_greeting(self, country_code: str, language: str) -> str:
        """
        Get localized greeting message for Aisle AI based on country/language
        """
        greetings = {
            'en': {
                'KE': "Karibu to AisleMarts Kenya! 🇰🇪 I'm Aisle, your shopping companion. Ready to discover amazing local and global products?",
                'NG': "Welcome to AisleMarts Nigeria! 🇳🇬 I'm Aisle, here to help you find the best deals across Africa and beyond.",
                'US': "Welcome to AisleMarts! 🇺🇸 I'm Aisle, your AI shopping guide ready to help you discover amazing products.",
                'GB': "Welcome to AisleMarts UK! 🇬🇧 I'm Aisle, delighted to help you explore our global marketplace.",
                'default': "Welcome to AisleMarts! 🌍 I'm Aisle, your global shopping companion ready to help you discover amazing products worldwide."
            },
            'ar': {
                'EG': "أهلاً بك في AisleMarts مصر! 🇪🇬 أنا Aisle، مساعدك الذكي للتسوق. مستعد لاكتشاف منتجات رائعة؟",
                'SA': "أهلاً وسهلاً بك في AisleMarts السعودية! 🇸🇦 أنا Aisle، هنا لمساعدتك في العثور على أفضل المنتجات.",
                'default': "أهلاً بك في AisleMarts! 🌍 أنا Aisle، مساعدك الذكي للتسوق العالمي."
            },
            'fr': {
                'FR': "Bienvenue sur AisleMarts France! 🇫🇷 Je suis Aisle, votre compagnon d'achat IA prêt à vous aider.",
                'default': "Bienvenue sur AisleMarts! 🌍 Je suis Aisle, votre guide d'achat mondial."
            },
            'es': {
                'ES': "¡Bienvenido a AisleMarts España! 🇪🇸 Soy Aisle, tu compañero de compras IA listo para ayudarte.",
                'MX': "¡Bienvenido a AisleMarts México! 🇲🇽 Soy Aisle, aquí para ayudarte a descubrir productos increíbles.",
                'default': "¡Bienvenido a AisleMarts! 🌍 Soy Aisle, tu guía de compras global."
            },
            'it': {
                'IT': "Benvenuto su AisleMarts Italia! 🇮🇹 Sono Aisle, il tuo compagno di shopping IA pronto ad aiutarti.",
                'default': "Benvenuto su AisleMarts! 🌍 Sono Aisle, la tua guida agli acquisti globale."
            },
            'tr': {
                'TR': "AisleMarts Türkiye'ye hoş geldiniz! 🇹🇷 Ben Aisle, alışveriş arkadaşınız. Harika ürünler keşfetmeye hazır mısınız?",
                'default': "AisleMarts'a hoş geldiniz! 🌍 Ben Aisle, küresel alışveriş rehberiniz."
            }
        }
        
        # Get language-specific greetings
        lang_greetings = greetings.get(language, greetings['en'])
        
        # Get country-specific or default greeting
        return lang_greetings.get(country_code, lang_greetings['default'])

# Initialize global instance
localization_service = LocalizationService()